src/lib/AST/ASTVisitor.cpp

```diff
@@ -2218,18 +2269,24 @@ class ASTVisitor
         I.Name = extractName(D);
         I.IsDirective = false;
 
-
         for (auto const* shadow : D->shadows())
         {
             NamedDecl* ND = shadow->getTargetDecl();
+
             SymbolID id;
-            extractSymbolID(ND, id);
-            I.UsingSymbols.emplace_back(id);
+            getDependencyID(ND, id);
+            if (id != SymbolID::invalid)
+            {
+                I.UsingSymbols.emplace_back(id);
 
-            // If this is a using declaration naming
-            // a previously undeclared namespace, traverse it.
-            if(ND->isFirstDecl()) {
-                traverseDecl(ND);
+                I.UsingName = std::make_unique<NameInfo>();
+                I.UsingName->id = id;
+                I.UsingName->Name = ND->getNameAsString();
+                if (auto const* parentContext = dyn_cast<NamedDecl>(ND->getDeclContext()))
```

Comment:
This doesn't build `NameInfo` in a fully recursive manner? It only works for a depth of 1. You should use `buildNameInfo(const NestedNameSpecifier*, ExtractMode)` to build the `NameInfo` for _just_ the _nested-name-specifier_ since we already store the terminal name of the using declaration in `Info::Name`. 

Also, the assignment of `UsingName` is in a loop, so it's overwritten for each `UsingShadowDecl`. You probably want to move this after line 2270 (`I.IsDirective = false;`). Something along the lines of `I.Qualifier = buildNameInfo(D->getQualifier());`.